File: services/portfolio-analytics/src/pfa/router.py
"""portfolio_analytics — HTTP router (same minimal pattern as S1/S2/S3/S5/S6/S8)."""

from __future__ import annotations

import re
from typing import Any, Callable, Dict, List, Optional, Tuple

Handler = Callable[..., Tuple[int, Dict[str, Any]]]


class Route:
    def __init__(self, method: str, pattern: str, handler: Handler) -> None:
        self.method = method.upper()
        parts = pattern.split("/")
        regex_parts: List[str] = []
        for part in parts:
            if part.startswith("{") and part.endswith("}"):
                regex_parts.append(f"(?P<{part[1:-1]}>[^/]+)")
            else:
                regex_parts.append(re.escape(part))
        self._regex = re.compile("^" + "/".join(regex_parts) + "$")
        self.handler = handler

    def match(self, method: str, path: str) -> Optional[Dict[str, str]]:
        if self.method != method.upper():
            return None
        m = self._regex.match(path)
        return m.groupdict() if m else None


class Router:
    def __init__(self) -> None:
        self._routes: List[Route] = []

    def add(self, method: str, pattern: str, handler: Handler) -> None:
        self._routes.append(Route(method, pattern, handler))

    def dispatch(self, method: str, path: str, query: Dict[str, str],
                 body: Optional[Dict[str, Any]] = None) -> Tuple[int, Dict[str, Any]]:
        for route in self._routes:
            params = route.match(method, path)
            if params is not None:
                try:
                    return route.handler(**params, _query=query, _body=body)
                except TypeError:
                    return route.handler(_query=query, _body=body)
        return 404, {
            "error": {
                "code": "PFA-404",
                "message": f"no route for {method} {path}",
                "service": "portfolio-analytics",
                "retryable": False,
                "context": {},
            }
        }


def build_router(controller) -> Router:
    router = Router()

    def healthz(_query=None, _body=None):
        return controller.healthz()

    def readyz(_query=None, _body=None):
        return controller.readyz()

    def pnl(_query=None, _body=None):
        return controller.pnl()

    def pnl_symbol(symbol: str, _query=None, _body=None):
        return controller.pnl_symbol(symbol)

    def metrics(_query=None, _body=None):
        return controller.metrics()

    def attribution(_query=None, _body=None):
        return controller.attribution()

    def var(_query=None, _body=None):
        q = _query or {}
        method = q.get("method")
        confidence_raw = q.get("confidence")
        confidence: Optional[float] = None
        if confidence_raw is not None:
            try:
                confidence = float(confidence_raw)
            except (TypeError, ValueError):
                confidence = None
        return controller.var(method=method, confidence=confidence)

    def history(_query=None, _body=None):
        q = _query or {}
        try:
            limit = int(q.get("limit", "100"))
        except (TypeError, ValueError):
            limit = 100
        return controller.history(limit=limit)

    def stats(_query=None, _body=None):
        return controller.stats()

    router.add("GET", "/healthz", healthz)
    router.add("GET", "/readyz", readyz)
    router.add("GET", "/pnl", pnl)
    router.add("GET", "/metrics", metrics)
    router.add("GET", "/attribution", attribution)
    router.add("GET", "/var", var)
    router.add("GET", "/history", history)
    router.add("GET", "/stats", stats)
    # /pnl/{symbol} is registered last so the literal segments above win.
    router.add("GET", "/pnl/{symbol}", pnl_symbol)
    return router

File: services/alerting-service/src/router.rs
//! alerting_service — HTTP route matching.
//!
//! A small method+path router with `{param}` segments, mirroring the other
//! services' pattern.  Routes are matched in registration order; the first
//! match wins and receives captured path parameters plus the parsed query.

use std::collections::HashMap;

/// The result of dispatching one request: HTTP status + JSON body bytes.
pub struct Response {
    pub status: u16,
    pub body: Vec<u8>,
}

impl Response {
    pub fn json(status: u16, value: &crate::json::Value) -> Self {
        Response {
            status,
            body: value.to_json().into_bytes(),
        }
    }

    pub fn error(err: &crate::errors::Error) -> Self {
        Response::json(err.http_status, &err.to_envelope())
    }
}

#[derive(Clone)]
enum Seg {
    Lit(String),
    Param(String),
}

/// A single registered route.
#[derive(Clone)]
struct Route {
    method: String,
    segments: Vec<Seg>,
    handler: crate::http::Handler,
}

impl Route {
    fn new(method: &str, pattern: &str, handler: crate::http::Handler) -> Self {
        let segments = pattern
            .trim_start_matches('/')
            .split('/')
            .map(|part| {
                if part.starts_with('{') && part.ends_with('}') {
                    Seg::Param(part[1..part.len() - 1].to_string())
                } else {
                    Seg::Lit(part.to_string())
                }
            })
            .collect();
        Route {
            method: method.to_uppercase(),
            segments,
            handler,
        }
    }

    /// Match a request path against this route (method already checked).
    fn match_path(&self, path: &str) -> Option<HashMap<String, String>> {
        let parts: Vec<&str> = path.trim_start_matches('/').split('/').collect();
        if parts.len() != self.segments.len() {
            return None;
        }
        let mut params = HashMap::new();
        for (part, seg) in parts.iter().zip(self.segments.iter()) {
            match seg {
                Seg::Lit(lit) => {
                    if lit != part {
                        return None;
                    }
                }
                Seg::Param(name) => {
                    params.insert(name.clone(), (*part).to_string());
                }
            }
        }
        Some(params)
    }
}

/// The router: an ordered list of routes.
#[derive(Clone)]
pub struct Router {
    routes: Vec<Route>,
}

impl Default for Router {
    fn default() -> Self {
        Router::new()
    }
}

impl Router {
    pub fn new() -> Self {
        Router { routes: Vec::new() }
    }

    pub fn add(&mut self, method: &str, pattern: &str, handler: crate::http::Handler) {
        self.routes.push(Route::new(method, pattern, handler));
    }

    /// Dispatch a request.  Returns the response to send back to the client.
    pub fn dispatch(
        &self,
        method: &str,
        path: &str,
        query: HashMap<String, String>,
        body: &str,
        manager: &crate::core::SharedManager,
    ) -> Response {
        let method = method.to_uppercase();
        for route in self.routes.iter() {
            if route.method != method {
                continue;
            }
            if let Some(params) = route.match_path(path) {
                return (route.handler)(RequestCtx {
                    method: method.clone(),
                    path,
                    params,
                    query,
                    body,
                    manager,
                });
            }
        }
        Response::error(&crate::errors::Error::not_found(&method, path))
    }

    /// Register the standard alerting-service routes.  The literal `/alerts/ack`
    /// and `/alerts/dispatch` are registered BEFORE the parameterized
    /// `/alerts/{id}` so the fixed segments win over the catch-all.
    pub fn with_default_routes() -> Self {
        let mut r = Router::new();
        use crate::http::{
            h_ack, h_alert, h_alerts, h_alerts_dispatch, h_alerts_list, h_healthz, h_readyz,
            h_resolve, h_stats,
        };
        r.add("GET", "/healthz", h_healthz);
        r.add("GET", "/readyz", h_readyz);
        r.add("POST", "/alerts", h_alerts);
        r.add("GET", "/alerts", h_alerts_list);
        r.add("GET", "/alerts/dispatch", h_alerts_dispatch);
        r.add("POST", "/alerts/ack", h_ack);
        r.add("POST", "/alerts/{id}/resolve", h_resolve);
        r.add("GET", "/alerts/{id}", h_alert);
        r.add("GET", "/stats", h_stats);
        r
    }
}

/// Context handed to each handler.
pub struct RequestCtx<'a> {
    pub method: String,
    pub path: &'a str,
    pub params: HashMap<String, String>,
    pub query: HashMap<String, String>,
    pub body: &'a str,
    pub manager: &'a crate::core::SharedManager,
}
